pkg/audit/log.go

```diff
@@ -0,0 +1,113 @@
+package audit
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/lyft/flyteadmin/pkg/common"
+	"github.com/lyft/flyteadmin/pkg/errors"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/lyft/flytestdlib/logger"
+	"google.golang.org/grpc/codes"
+)
+
+type LogBuilder interface {
+	WithAuthenticatedCtx(ctx context.Context) LogBuilder
+	WithRequest(method string, parameters map[string]string, mode admin.Request_Mode, requestedAt time.Time) LogBuilder
+	WithResponse(sentAt time.Time, err error) LogBuilder
+	Log(ctx context.Context)
+}
+
+var jsonPbMarshaler = jsonpb.Marshaler{}
+
+type logBuilder struct {
+	auditLog admin.AuditLog
+	readOnly bool
+}
+
+func (b *logBuilder) WithAuthenticatedCtx(ctx context.Context) LogBuilder {
+	clientMeta := ctx.Value(common.AuditFieldsContextKey)
+	switch clientMeta.(type) {
+	case AuthenticatedClientMeta:
+		tokenIssuedAt, err := ptypes.TimestampProto(clientMeta.(AuthenticatedClientMeta).TokenIssuedAt)
+		if err != nil {
+			logger.Warningf(ctx, "Failed to convert authenticated TokenIssuedAt to timestamp proto: %v", err)
+		}
+		b.auditLog.Principal = &admin.Principal{
+			Subject:       ctx.Value(common.PrincipalContextKey).(string),
+			TokenIssuedAt: tokenIssuedAt,
+		}
+		if len(clientMeta.(AuthenticatedClientMeta).ClientIds) > 0 {
+			b.auditLog.Principal.ClientId = clientMeta.(AuthenticatedClientMeta).ClientIds[0]
```

Comment:
will we ever store more than the first one?

Reply:
I'm not sure - is the convention ever to have multiple client ids? i can make this an array/slice, what do you think?